[SN-142]

jbkkkkk opened the pull request.

jbkkkkk committed:
[SN-142]
Добавлена система жалоб

c23p1 commented on src/Application/Interfaces/Repositories/Admin/IReportCategoriesRepository.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public Task<ReportCategory?> GetByIdAsync(int id);
+    public Task CreateAsync(ReportCategory reportCategory);
+    public Task DeleteByIdAsync(int id);
```
Не должно быть возможности удаления категорий

c23p1 commented on src/Application/DTO/ReportDTO.cs:
```diff
@@ -0,0 +1,14 @@
+public record ReportDTO
+{
+    [Required] 
+    public int TargetId { get; init; }
```
Тип данных целых чисел без null - int? с [Required]

c23p1 commented on src/Application/DTO/ReportDTO.cs:
```diff
@@ -0,0 +1,14 @@
+    [Required] 
+    public int TargetId { get; init; }
+    [Required]
+    public int TargetEntityId { get; init; }
```
Тип данных целых чисел без null - int? с [Required]

c23p1 commented on src/Application/DTO/ReportDTO.cs:
```diff
@@ -0,0 +1,14 @@
+    [Required]
+    public int TargetEntityId { get; init; }
+    [Required]
+    public int CategoryId { get; init; }
```
Тип данных целых чисел без null - int? с [Required]

c23p1 commented on src/Admin.API/Controllers/ReportController.cs:
```diff
@@ -0,0 +1,42 @@
+
+    [HttpGet]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    public async Task<ActionResult<ICollection<Report>>> GetReports() => Ok(await _reportsQueryService.GetAllAsync());
```
Нейминг - GetAll

c23p1 commented on src/Admin.API/Controllers/ReportController.cs:
```diff
@@ -0,0 +1,42 @@
+    [HttpGet("{id:int}")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<ReportDTO>> GetReport(int id) => Ok(await _reportsQueryService.GetByIdAsync(id));
```
Нейминг - GetById

c23p1 commented on src/Application/Services/ReportsService.cs:
```diff
@@ -0,0 +1,80 @@
+
+        switch (reportDTO.TargetEntityId)
+        {
+            case 1:
```
Magic number 1

c23p1 commented on src/Application/Services/ReportsService.cs:
```diff
@@ -0,0 +1,80 @@
+                    throw new NotFoundException(nameof(PersonModel), reportDTO.TargetId);
+                report.TargetPersonId = reportDTO.TargetId;
+                break;
+            case 2:
```
Magic number 2

c23p1 commented on src/Application/Services/ReportsService.cs:
```diff
@@ -0,0 +1,80 @@
+                    throw new NotFoundException(nameof(ClubModel), reportDTO.TargetId);
+                report.TargetClubId = reportDTO.TargetId;
+                break;
+            case 3:
```
Magic number 3

c23p1 commented on src/Application/Services/ReportsService.cs:
```diff
@@ -0,0 +1,80 @@
+                report.TargetPostId = reportDTO.TargetId;
+                break;
+            default:
+                throw new ArgumentException("Некорректная сущность для репорта");
```
- Исправить на "Указана некорректная сущность при создании жалобы"

jbkkkkk committed:
[SN-142]
Изменены названия методов в ReportController, создан enum ReportTargetEntities, удалены лишние атрибуты Required, удален метод DeleteByIdAsync из ReportCategoriesRepository, изменено сообщение при указании некорректной сущности для жалобы

jbkkkkk committed:
[SN-142]
Внесены незначительные изменения

jbkkkkk committed:
[SN-142]
Добавлена возможность получения жалоб по сущности

c23p1 commented on src/Infrastructure/Services/Queries/ReportsQueryService.cs:
```diff
@@ -24,4 +25,17 @@ public async Task<Report> GetByIdAsync(int id)
+                return await _reports.Where(r => r.TargetPersonId != null).ToListAsync();
+            case ReportTargetEntities.Club:
+                return await _reports.Where(r => r.TargetClubId != null).ToListAsync();
+            default:
```
Случай по-умолчанию не должен быть предусмотрен при использовании enum

c23p1 commented on src/Admin.API/Controllers/ReportController.cs:
```diff
@@ -39,4 +40,9 @@ public async Task<IActionResult> Create(ReportDTO reportDTO)
+
+    [HttpGet("getByEntityId")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status409Conflict)]
```
Conflict исключён

c23p1 commented on src/Application/Interfaces/Queries/IReportsQueryService.cs:
```diff
@@ -6,4 +6,5 @@ public interface IReportsQueryService
 {
     public Task<ICollection<Report>> GetAllAsync();
     public Task<Report> GetByIdAsync(int id);
+    public Task<ICollection<Report>> GetByEntityIdAsync(ReportTargetEntities entity);
```
GetByTargetEntityAsync

c23p1 commented on src/Admin.API/Controllers/ReportController.cs:
```diff
@@ -39,4 +40,9 @@ public async Task<IActionResult> Create(ReportDTO reportDTO)
+    [HttpGet("getByEntityId")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status409Conflict)]
+    public async Task<ActionResult<ICollection<Report>>> GetByEntityId(ReportTargetEntities entity) => Ok(await _reportsQueryService.GetByEntityIdAsync(entity));
```
GetByTargetEntityAsync

jbkkkkk committed:
[SN-142]
GetByEntityId заменен на GetByTargetEntityId, незначительные изменения
